File: helfomat-core/src/main/java/de/helfenkannjeder/helfomat/core/organization/event/OrganizationEditAddVolunteerEvent.java
package de.helfenkannjeder.helfomat.core.organization.event;

import de.helfenkannjeder.helfomat.core.organization.Organization;
import de.helfenkannjeder.helfomat.core.organization.OrganizationEventVisitor;
import de.helfenkannjeder.helfomat.core.organization.OrganizationId;
import de.helfenkannjeder.helfomat.core.organization.Volunteer;

/**
 * @author Valentin Zickner
 */
@SuppressWarnings({"WeakerAccess", "CanBeFinal"})
public class OrganizationEditAddVolunteerEvent extends OrganizationEditEvent {
    private int index;
    private Volunteer volunteer;

    protected OrganizationEditAddVolunteerEvent() {
    }

    public OrganizationEditAddVolunteerEvent(OrganizationId organizationId, int index, Volunteer volunteer) {
        super(organizationId);
        this.index = index;
        this.volunteer = volunteer;
    }

    public int getIndex() {
        return index;
    }

    public Volunteer getVolunteer() {
        return volunteer;
    }

    @Override
    public Organization.Builder applyOnOrganizationBuilder(Organization.Builder organization) {
        return organization.addVolunteer(index, volunteer);
    }

    @Override
    public <T> T visit(OrganizationEventVisitor<T> visitor) {
        return visitor.visit(this);
    }
}

File: helfomat-core/src/main/kotlin/de/helfenkannjeder/helfomat/core/organization/event/OrganizationEditAddVolunteerEvent.kt
package de.helfenkannjeder.helfomat.core.organization.event

import de.helfenkannjeder.helfomat.core.organization.Organization
import de.helfenkannjeder.helfomat.core.organization.OrganizationEventVisitor
import de.helfenkannjeder.helfomat.core.organization.OrganizationId
import de.helfenkannjeder.helfomat.core.organization.Volunteer

/**
 * @author Valentin Zickner
 */
data class OrganizationEditAddVolunteerEvent(
    override val organizationId: OrganizationId,
    val index: Int,
    val volunteer: Volunteer
) : OrganizationEditEvent(organizationId) {

    override fun applyOnOrganizationBuilder(organizationBuilder: Organization.Builder?): Organization.Builder? {
        if (organizationBuilder?.volunteers != null) {
            if (organizationBuilder.volunteers.size > index) {
                organizationBuilder.volunteers.add(index, volunteer)
            } else {
                organizationBuilder.volunteers.add(volunteer)
            }
        }
        return organizationBuilder
    }

    override fun <T> visit(visitor: OrganizationEventVisitor<T>): T {
        return visitor.visit(this)
    }

}